File: app/src/test/java/info/nightscout/utils/BolusWizardTest.java
package info.nightscout.utils;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.data.GlucoseStatus;
import info.nightscout.androidaps.data.IobTotal;
import info.nightscout.androidaps.data.Profile;
import info.nightscout.androidaps.interfaces.PumpInterface;
import info.nightscout.androidaps.plugins.ConfigBuilder.ConfigBuilderPlugin;
import info.nightscout.androidaps.plugins.PumpMDI.MDIPlugin;
import info.nightscout.androidaps.plugins.Treatments.TreatmentsPlugin;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * Created by kuchjir on 12/12/2017.
 */
@RunWith(PowerMockRunner.class)
@PrepareForTest( { MainApp.class, GlucoseStatus.class, ConfigBuilderPlugin.class, TreatmentsPlugin.class })
public class BolusWizardTest {
    private static final double PUMP_BOLUS_STEP = 0.1;

    @Test
    /** Should calculate the same bolus when different blood glucose but both in target range */
    public void shuldCalculateTheSameBolusWhenBGsInRange() throws Exception {
        BolusWizard bw = new BolusWizard();
        Profile profile = setupProfile(4d, 8d, 20d, 12d);

        Double bolusForBg42 = bw.doCalc(profile, null, 20, 0.0,4.2, 0d, 100d, true, true, false, false);
        Double bolusForBg54 = bw.doCalc(profile, null, 20, 0.0,5.4, 0d, 100d, true, true, false, false);
        Assert.assertEquals(bolusForBg42, bolusForBg54);
    }

    @Test
    public void shuldCalculateHigherBolusWhenHighBG() throws Exception {
        BolusWizard bw = new BolusWizard();
        Profile profile = setupProfile(4d, 8d, 20d, 12d);

        Double bolusForHighBg = bw.doCalc(profile, null, 20, 0d,9.8, 0d, 100d, true, true, false, false);
        Double bolusForBgInRange = bw.doCalc(profile, null, 20, 0.0,5.4, 0d, 100d, true, true, false, false);
        Assert.assertTrue(bolusForHighBg > bolusForBgInRange);
    }

    @Test
    public void shuldCalculateLowerBolusWhenLowBG() throws Exception {
        BolusWizard bw = new BolusWizard();
        Profile profile = setupProfile(4d, 8d, 20d, 12d);

        Double bolusForLowBg = bw.doCalc(profile, null, 20, 0d,3.6, 0d, 100d, true, true, false, false);
        Double bolusForBgInRange = bw.doCalc(profile, null, 20, 0.0,5.4, 0d, 100d, true, true, false, false);
        Assert.assertTrue(bolusForLowBg < bolusForBgInRange);
    }

    private Profile setupProfile(Double targetLow, Double targetHigh, Double insulinSensitivityFactor, Double insulinToCarbRatio) {
        Profile profile = mock(Profile.class);
        when(profile.getTargetLow()).thenReturn(targetLow);
        when(profile.getTargetHigh()).thenReturn(targetHigh);
        when(profile.getIsf()).thenReturn(insulinSensitivityFactor);
        when(profile.getIc()).thenReturn(insulinToCarbRatio);

        PowerMockito.mockStatic(GlucoseStatus.class);
        when(GlucoseStatus.getGlucoseStatusData()).thenReturn(null);

        PowerMockito.mockStatic(TreatmentsPlugin.class);
        TreatmentsPlugin treatment = mock(TreatmentsPlugin.class);
        IobTotal iobTotalZero = new IobTotal(System.currentTimeMillis());
        when(treatment.getLastCalculationTreatments()).thenReturn(iobTotalZero);
        when(treatment.getLastCalculationTempBasals()).thenReturn(iobTotalZero);
        PowerMockito.mockStatic(MainApp.class);
        when(TreatmentsPlugin.getPlugin()).thenReturn(treatment);

        PowerMockito.mockStatic(ConfigBuilderPlugin.class);
        PumpInterface pump = MDIPlugin.getPlugin();
        pump.getPumpDescription().bolusStep = PUMP_BOLUS_STEP;
        when(ConfigBuilderPlugin.getActivePump()).thenReturn(pump);

        return profile;
    }
}

File: app/src/test/java/info/nightscout/androidaps/utils/wizard/BolusWizardTest.kt
package info.nightscout.androidaps.utils.wizard

import android.content.Context
import dagger.android.AndroidInjector
import dagger.android.HasAndroidInjector
import info.nightscout.androidaps.TestBase
import info.nightscout.androidaps.Constants
import info.nightscout.androidaps.data.IobTotal
import info.nightscout.androidaps.data.Profile
import info.nightscout.androidaps.interfaces.ActivePluginProvider
import info.nightscout.androidaps.interfaces.CommandQueueProvider
import info.nightscout.androidaps.interfaces.Constraint
import info.nightscout.androidaps.interfaces.PumpDescription
import info.nightscout.androidaps.plugins.aps.loop.LoopPlugin
import info.nightscout.androidaps.plugins.bus.RxBusWrapper
import info.nightscout.androidaps.plugins.configBuilder.ConstraintChecker
import info.nightscout.androidaps.interfaces.ProfileFunction
import info.nightscout.androidaps.plugins.iob.iobCobCalculator.GlucoseStatus
import info.nightscout.androidaps.plugins.iob.iobCobCalculator.IobCobCalculatorPlugin
import info.nightscout.androidaps.plugins.pump.virtual.VirtualPumpPlugin
import info.nightscout.androidaps.plugins.treatments.TreatmentsPlugin
import info.nightscout.androidaps.utils.resources.ResourceHelper
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.Mockito.`when`
import org.mockito.invocation.InvocationOnMock
import org.powermock.core.classloader.annotations.PrepareForTest
import org.powermock.modules.junit4.PowerMockRunner

@RunWith(PowerMockRunner::class)
@PrepareForTest(ConstraintChecker::class, VirtualPumpPlugin::class)
class BolusWizardTest : TestBase() {

    private val PUMP_BOLUS_STEP = 0.1

    @Mock lateinit var resourceHelper: ResourceHelper
    @Mock lateinit var profileFunction: ProfileFunction
    @Mock lateinit var constraintChecker: ConstraintChecker
    @Mock lateinit var context: Context
    @Mock lateinit var activePlugin: ActivePluginProvider
    @Mock lateinit var commandQueue: CommandQueueProvider
    @Mock lateinit var loopPlugin: LoopPlugin
    @Mock lateinit var iobCobCalculatorPlugin: IobCobCalculatorPlugin
    @Mock lateinit var treatmentsPlugin: TreatmentsPlugin
    @Mock lateinit var virtualPumpPlugin: VirtualPumpPlugin

    val injector = HasAndroidInjector {
        AndroidInjector {
            if (it is BolusWizard) {
                it.aapsLogger = aapsLogger
                it.resourceHelper = resourceHelper
                it.rxBus = RxBusWrapper()
                it.profileFunction = profileFunction
                it.constraintChecker = constraintChecker
                it.activePlugin = activePlugin
                it.commandQueue = commandQueue
                it.loopPlugin = loopPlugin
                it.iobCobCalculatorPlugin = iobCobCalculatorPlugin
            }
            if (it is GlucoseStatus) {
                it.aapsLogger = aapsLogger
                it.iobCobCalculatorPlugin = iobCobCalculatorPlugin
            }
        }
    }

    private fun setupProfile(targetLow: Double, targetHigh: Double, insulinSensitivityFactor: Double, insulinToCarbRatio: Double): Profile {
        val profile = Mockito.mock(Profile::class.java)
        `when`(profile.targetLowMgdl).thenReturn(targetLow)
        `when`(profile.targetHighMgdl).thenReturn(targetHigh)
        `when`(profile.isfMgdl).thenReturn(insulinSensitivityFactor)
        `when`(profile.ic).thenReturn(insulinToCarbRatio)

        `when`(profileFunction.getUnits()).thenReturn(Constants.MGDL)
        `when`(iobCobCalculatorPlugin.dataLock).thenReturn(Unit)
        `when`(activePlugin.activeTreatments).thenReturn(treatmentsPlugin)
        `when`(treatmentsPlugin.lastCalculationTreatments).thenReturn(IobTotal(System.currentTimeMillis()))
        `when`(treatmentsPlugin.lastCalculationTempBasals).thenReturn(IobTotal(System.currentTimeMillis()))
        `when`(activePlugin.activePump).thenReturn(virtualPumpPlugin)
        val pumpDescription = PumpDescription()
        pumpDescription.bolusStep = PUMP_BOLUS_STEP
        `when`(virtualPumpPlugin.pumpDescription).thenReturn(pumpDescription)

        Mockito.doAnswer { invocation: InvocationOnMock ->
            invocation.getArgument<Constraint<Double>>(0)
        }.`when`(constraintChecker).applyBolusConstraints(anyObject())
        return profile
    }

    @Test
        /** Should calculate the same bolus when different blood glucose but both in target range  */
    fun shouldCalculateTheSameBolusWhenBGsInRange() {
        val profile = setupProfile(4.0, 8.0, 20.0, 12.0)
        var bw = BolusWizard(injector).doCalc(profile, "", null, 20, 0.0, 4.2, 0.0, 100.0, true, true, true, true, false, false, false, false)
        val bolusForBg42 = bw.calculatedTotalInsulin
        bw = BolusWizard(injector).doCalc(profile, "", null, 20, 0.0, 5.4, 0.0, 100.0, true, true, true, true, false, false, false, false)
        val bolusForBg54 = bw.calculatedTotalInsulin
        Assert.assertEquals(bolusForBg42, bolusForBg54, 0.01)
    }

    @Test
    fun shouldCalculateHigherBolusWhenHighBG() {
        val profile = setupProfile(4.0, 8.0, 20.0, 12.0)
        var bw = BolusWizard(injector).doCalc(profile, "", null, 20, 0.0, 9.8, 0.0, 100.0, true, true, true, true, false, false, false, false)
        val bolusForHighBg = bw.calculatedTotalInsulin
        bw = BolusWizard(injector).doCalc(profile, "", null, 20, 0.0, 5.4, 0.0, 100.0, true, true, true, true, false, false, false, false)
        val bolusForBgInRange = bw.calculatedTotalInsulin
        Assert.assertTrue(bolusForHighBg > bolusForBgInRange)
    }

    @Test
    fun shouldCalculateLowerBolusWhenLowBG() {
        val profile = setupProfile(4.0, 8.0, 20.0, 12.0)
        var bw = BolusWizard(injector).doCalc(profile, "", null, 20, 0.0, 3.6, 0.0, 100.0, true, true, true, true, false, false, false, false)
        val bolusForLowBg = bw.calculatedTotalInsulin
        bw = BolusWizard(injector).doCalc(profile, "", null, 20, 0.0, 5.4, 0.0, 100.0, true, true, true, true, false, false, false, false)
        val bolusForBgInRange = bw.calculatedTotalInsulin
        Assert.assertTrue(bolusForLowBg < bolusForBgInRange)
    }
}